src/ucp/core/ucp_ep.c

```diff
@@ -2670,37 +2670,74 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
     }
 }
 
+static int ucp_ep_req_unrelease(ucp_request_t *req)
+{
+    int req_release = req->flags & UCP_REQUEST_FLAG_RELEASED;
+
+    /* Mark release to be not released from the ucp_request_complete() procedure
+     * to be able do some actions with this request (e.g. check and release
+     * request ID) after completing an operation.
```

Comment:
i think it's better to release the request id for each protocol explicitly. after we find which protocol is it according to req flags, we can decide whether need to release the id or not. and potentially unite code with normal-flow completion and id release.

Reply:
do you mean to release ID from protocol completion callback?
but it is not possible sometimes, because some of them, e.g. RNDV PUT rely on the fact that request ID will be released upon receiving RTR packet, since ID is no longer needed

Reply:
I mean from the if/else/else ... code below
is it possible some protocols release the id inside the completion callback?